File: android/MyNotes/app/src/main/java/io/qzz/superstar1014/mynotes/HermesChatClient.java
package io.qzz.superstar1014.mynotes;

import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;

/** One process-wide WebSocket that multiplexes all configured Hermes profiles by spaceId. */
final class HermesChatClient {
  static final String RELAY_URL = "wss://h.superstar1014.qzz.io/api/hermes-chat/ws";
  private static final long HANDSHAKE_TIMEOUT_MS = 20_000L;

  interface Listener {
    void onReady();
    void onMessage(
      String spaceId,
      HermesChatCrypto.ChatMessage message,
      String sessionToken
    );
    void onAcknowledged(
      String spaceId,
      String messageId,
      long sequence,
      String sessionToken
    );
    void onActionAcknowledged(
      String spaceId,
      String actionId,
      boolean delivered,
      String sessionToken
    );
    void onTyping(
      String spaceId,
      boolean active,
      boolean terminal,
      String sessionToken
    );
    void onProfileError(String spaceId, String reason, String sessionToken);
    void onDisconnected(String reason);
  }

  private final Listener listener;
  private final OkHttpClient http;
  private final Handler handler = new Handler(Looper.getMainLooper());
  private final Object socketStateLock = new Object();
  private final Object profileStateLock = new Object();
  private final Map<String, ProfileState> profiles = new ConcurrentHashMap<>();
  private volatile WebSocket socket;
  private volatile boolean closing;
  private Runnable handshakeTimeoutTask;
  private boolean handshakePending;

  HermesChatClient(Listener listener) {
    this.listener = listener;
    http = new OkHttpClient.Builder()
      .connectTimeout(20, TimeUnit.SECONDS)
      .readTimeout(0, TimeUnit.MILLISECONDS)
      .pingInterval(25, TimeUnit.SECONDS)
      .retryOnConnectionFailure(true)
      .build();
  }

  void configureProfile(
    String spaceId,
    HermesChatCrypto crypto,
    String keyFingerprint,
    long initialSequence
  ) {
    if (initialSequence < 0) throw new IllegalArgumentException("聊天恢复序号无效。");
    synchronized (profileStateLock) {
      profiles.put(spaceId, new ProfileState(crypto, keyFingerprint, initialSequence));
    }
  }

  void retainProfiles(Set<String> spaceIds) {
    synchronized (profileStateLock) {
      profiles.keySet().removeIf(spaceId -> !spaceIds.contains(spaceId));
    }
  }

  void connect(Models.HermesChatMultiplexTicket ticket) {
    close();
    closing = false;
    Request request = new Request.Builder()
      .url(RELAY_URL)
      .header("Authorization", "Bearer " + ticket.ticket)
      .header("X-Hermes-Role", "client")
      .header("X-Hermes-Mode", "multiplex")
      .build();
    Map<String, String> sessionTokens = new ConcurrentHashMap<>();
    synchronized (profileStateLock) {
      for (String spaceId : ticket.spaceIds) {
        ProfileState profile = profiles.get(spaceId);
        if (profile != null) sessionTokens.put(spaceId, profile.keyFingerprint);
      }
    }
    WebSocket nextSocket = http.newWebSocket(
      request,
      new SocketListener(List.copyOf(ticket.spaceIds), Map.copyOf(sessionTokens))
    );
    synchronized (socketStateLock) {
      socket = nextSocket;
      scheduleHandshakeTimeoutLocked(nextSocket);
    }
  }

  HermesChatCrypto.ChatMessage sendMessage(
    String spaceId,
    String text,
    JSONArray attachments,
    String expectedSessionToken
  ) throws Exception {
    synchronized (profileStateLock) {
      ProfileState profile = requireProfile(spaceId);
      if (expectedSessionToken != null
          && !profile.keyFingerprint.equals(expectedSessionToken)) {
        throw new IllegalStateException("Hermes 聊天密钥已更改，请重新发送附件。");
      }
      String id = UUID.randomUUID().toString();
      long sentAt = System.currentTimeMillis();
      JSONObject envelope = profile.crypto.encryptMessage(
        "client",
        text,
        attachments,
        id,
        sentAt
      );
      if (!send(spaceId, envelope)) throw new IllegalStateException("聊天连接尚未就绪。");
      return new HermesChatCrypto.ChatMessage(
        id,
        "client",
        sentAt,
        0,
        text == null ? "" : text,
        attachments == null ? new JSONArray() : attachments
      );
    }
  }

  String sendInteractionAction(
    String spaceId,
    String promptId,
    String optionId,
    String actionId
  ) throws Exception {
    synchronized (profileStateLock) {
      ProfileState profile = requireProfile(spaceId);
      JSONObject message = profile.crypto.encryptInteractionResponse(
        promptId,
        optionId,
        actionId,
        System.currentTimeMillis()
      );
      JSONObject frame = new JSONObject()
        .put("v", 1)
        .put("type", "action")
        .put("message", message);
      if (!send(spaceId, frame)) throw new IllegalStateException("聊天连接尚未就绪。");
      return actionId;
    }
  }

  void sendTyping(String spaceId, boolean active) {
    ProfileState profile = profiles.get(spaceId);
    if (profile == null || (profile.lastTypingActive != null
        && profile.lastTypingActive == active)) return;
    try {
      boolean sent = send(spaceId, new JSONObject()
        .put("v", 1)
        .put("type", "typing")
        .put("active", active));
      if (sent) profile.lastTypingActive = active;
    } catch (Exception ignored) {
      // Typing state is ephemeral and must not interrupt message entry.
    }
  }

  boolean isConnected() {
    return socket != null && !closing;
  }

  void close() {
    closing = true;
    for (ProfileState profile : profiles.values()) profile.lastTypingActive = null;
    WebSocket current;
    synchronized (socketStateLock) {
      cancelHandshakeTimeoutLocked();
      current = socket;
      socket = null;
    }
    if (current != null) current.close(1000, "Android chat closed");
  }

  void shutdown() {
    close();
    http.dispatcher().executorService().shutdown();
    http.connectionPool().evictAll();
  }

  private ProfileState requireProfile(String spaceId) {
    ProfileState profile = profiles.get(spaceId);
    if (profile == null) throw new IllegalStateException("该 Hermes profile 尚未配置聊天密钥。");
    return profile;
  }

  private boolean send(String spaceId, JSONObject frame) throws Exception {
    WebSocket current = socket;
    if (current == null) return false;
    frame.put("spaceId", spaceId);
    return current.send(frame.toString());
  }

  private void scheduleHandshakeTimeoutLocked(WebSocket expectedSocket) {
    cancelHandshakeTimeoutLocked();
    handshakePending = true;
    handshakeTimeoutTask = () -> failHandshakeIfPending(expectedSocket);
    handler.postDelayed(handshakeTimeoutTask, HANDSHAKE_TIMEOUT_MS);
  }

  private boolean cancelHandshakeTimeout(WebSocket expectedSocket) {
    synchronized (socketStateLock) {
      if (socket != expectedSocket || closing || !handshakePending) return false;
      cancelHandshakeTimeoutLocked();
      return true;
    }
  }

  private void cancelHandshakeTimeoutLocked() {
    handshakePending = false;
    if (handshakeTimeoutTask == null) return;
    handler.removeCallbacks(handshakeTimeoutTask);
    handshakeTimeoutTask = null;
  }

  private void failHandshakeIfPending(WebSocket expectedSocket) {
    synchronized (socketStateLock) {
      if (socket != expectedSocket || closing || !handshakePending) return;
      handshakePending = false;
      handshakeTimeoutTask = null;
      socket = null;
    }
    for (ProfileState profile : profiles.values()) profile.lastTypingActive = null;
    expectedSocket.cancel();
    listener.onDisconnected("WebSocket 握手超时，正在自动重试。");
  }

  private final class SocketListener extends WebSocketListener {
    private final List<String> ticketSpaces;
    private final Map<String, String> sessionTokens;

    SocketListener(List<String> ticketSpaces, Map<String, String> sessionTokens) {
      this.ticketSpaces = ticketSpaces;
      this.sessionTokens = sessionTokens;
    }

    @Override
    public void onOpen(WebSocket webSocket, Response response) {
      if (webSocket != socket) {
        webSocket.close(1000, "Replaced by a newer chat connection");
        return;
      }
      for (ProfileState profile : profiles.values()) profile.lastTypingActive = null;
    }

    @Override
    public void onMessage(WebSocket webSocket, String text) {
      if (webSocket != socket) return;
      String spaceId = "";
      String sessionToken = "";
      try {
        JSONObject frame = new JSONObject(text);
        String type = frame.optString("type");
        if ("ready".equals(type)) {
          if (!frame.optBoolean("multiplex")) {
            throw new IllegalArgumentException("服务器未启用多会话连接。");
          }
          if (!cancelHandshakeTimeout(webSocket)) return;
          for (String ticketSpace : ticketSpaces) sendResume(ticketSpace);
          listener.onReady();
          return;
        }
        if ("pong".equals(type)) return;
        spaceId = frame.getString("spaceId");
        sessionToken = sessionTokens.get(spaceId);
        if (sessionToken == null) return;
        ProfileState profile = requireProfile(spaceId);
        if (!profile.keyFingerprint.equals(sessionToken)) return;
        if ("message".equals(type)) {
          long sequence = frame.optLong("seq");
          HermesChatCrypto.ChatMessage message = profile.crypto.decryptMessage(
            frame.getJSONObject("message"),
            sequence
          );
          profile.lastSequence = Math.max(profile.lastSequence, sequence);
          listener.onMessage(spaceId, message, sessionToken);
          return;
        }
        if ("edit".equals(type)) {
          long sequence = frame.optLong("seq");
          HermesChatCrypto.ChatMessage message = profile.crypto.decryptMessage(
            frame.getJSONObject("message"),
            sequence
          );
          if (message.replaceSequence != frame.getLong("targetSeq")
              || message.finalUpdate != frame.getBoolean("final")) {
            throw new IllegalArgumentException("流式消息替换元数据校验失败。");
          }
          if (sequence > 0) profile.lastSequence = Math.max(profile.lastSequence, sequence);
          listener.onMessage(spaceId, message, sessionToken);
          return;
        }
        if ("resume_complete".equals(type)) {
          if (frame.optBoolean("hasMore")) sendResume(spaceId);
          return;
        }
        if ("ack".equals(type)) {
          if (!frame.optBoolean("durable", true)) {
            listener.onActionAcknowledged(
              spaceId,
              frame.optString("id"),
              frame.optBoolean("delivered"),
              sessionToken
            );
            return;
          }
          long sequence = frame.optLong("seq");
          if (sequence > 0) profile.lastSequence = Math.max(profile.lastSequence, sequence);
          listener.onAcknowledged(
            spaceId,
            frame.optString("id"),
            sequence,
            sessionToken
          );
          return;
        }
        if ("typing".equals(type) && "agent".equals(frame.optString("sender"))) {
          boolean active = frame.optBoolean("active");
          boolean terminal = !active
            && (!frame.has("terminal") || frame.optBoolean("terminal"));
          listener.onTyping(spaceId, active, terminal, sessionToken);
          return;
        }
        if ("error".equals(type)) {
          listener.onProfileError(
            spaceId,
            frame.optString("message", "Cloudflare 拒绝了聊天消息。"),
            sessionToken
          );
        }
      } catch (Exception error) {
        String reason = "无法解析加密聊天消息：" + error.getMessage();
        if (spaceId.isEmpty()) listener.onDisconnected(reason);
        else listener.onProfileError(spaceId, reason, sessionToken);
      }
    }

    private void sendResume(String spaceId) throws Exception {
      ProfileState profile = requireProfile(spaceId);
      send(spaceId, new JSONObject()
        .put("v", 1)
        .put("type", "resume")
        .put("afterSeq", profile.lastSequence));
    }

    @Override
    public void onClosing(WebSocket webSocket, int code, String reason) {
      webSocket.close(code, reason);
    }

    @Override
    public void onClosed(WebSocket webSocket, int code, String reason) {
      synchronized (socketStateLock) {
        if (webSocket != socket) return;
        cancelHandshakeTimeoutLocked();
        socket = null;
      }
      for (ProfileState profile : profiles.values()) profile.lastTypingActive = null;
      if (!closing) listener.onDisconnected(reason.isEmpty() ? "聊天连接已断开。" : reason);
    }

    @Override
    public void onFailure(WebSocket webSocket, Throwable error, Response response) {
      synchronized (socketStateLock) {
        if (webSocket != socket) return;
        cancelHandshakeTimeoutLocked();
        socket = null;
      }
      for (ProfileState profile : profiles.values()) profile.lastTypingActive = null;
      if (!closing) {
        String message = error.getMessage();
        listener.onDisconnected(message == null ? "无法连接 Hermes。" : message);
      }
    }
  }

  private static final class ProfileState {
    final HermesChatCrypto crypto;
    final String keyFingerprint;
    volatile long lastSequence;
    volatile Boolean lastTypingActive;

    ProfileState(
      HermesChatCrypto crypto,
      String keyFingerprint,
      long initialSequence
    ) {
      this.crypto = crypto;
      this.keyFingerprint = keyFingerprint;
      this.lastSequence = initialSequence;
    }
  }
}

File: macos/AIBuildNotes/Sources/NotesCore/HermesChatClient.swift
import Foundation

public struct HermesChatClientProfile: Sendable {
  public let spaceID: String
  public let crypto: HermesChatCrypto
  public let lastSequence: Int64
  public let resumeSinceMilliseconds: Int64?

  public init(
    spaceID: String,
    crypto: HermesChatCrypto,
    lastSequence: Int64,
    resumeSinceMilliseconds: Int64? = nil
  ) {
    self.spaceID = spaceID
    self.crypto = crypto
    self.lastSequence = max(0, lastSequence)
    if let resumeSinceMilliseconds, resumeSinceMilliseconds > 0 {
      self.resumeSinceMilliseconds = resumeSinceMilliseconds
    } else {
      self.resumeSinceMilliseconds = nil
    }
  }
}

public enum HermesChatClientEvent: Sendable {
  case ready
  case message(spaceID: String, message: HermesChatMessage)
  case acknowledged(spaceID: String, messageID: String, sequence: Int64)
  case resumed(spaceID: String, latestSequence: Int64)
  case actionAcknowledged(spaceID: String, actionID: String, delivered: Bool)
  case typing(spaceID: String, active: Bool)
  case profileError(spaceID: String, reason: String)
  case disconnected(reason: String)
}

public actor HermesChatClient {
  public static let productionRelayURL = URL(
    string: "wss://h.superstar1014.qzz.io/api/hermes-chat/ws"
  )!
  static let handshakeTimeoutMessage = "WebSocket 握手超时，正在自动重试。"

  public nonisolated let events: AsyncStream<HermesChatClientEvent>

  private let continuation: AsyncStream<HermesChatClientEvent>.Continuation
  private let relayURL: URL
  private let session: URLSession
  private let handshakeTimeout: Duration
  private var socket: URLSessionWebSocketTask?
  private var profiles: [String: ProfileState] = [:]
  private var receiveTask: Task<Void, Never>?
  private var pingTask: Task<Void, Never>?
  private var handshakeTask: Task<Void, Never>?
  private var generation = 0
  private var closing = false
  private var ready = false

  public init(
    relayURL: URL = productionRelayURL,
    session: URLSession? = nil,
    handshakeTimeout: Duration = .seconds(15)
  ) {
    self.relayURL = relayURL
    self.handshakeTimeout = handshakeTimeout
    if let session {
      self.session = session
    } else {
      let configuration = URLSessionConfiguration.default
      configuration.waitsForConnectivity = true
      configuration.timeoutIntervalForRequest = 30
      configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
      self.session = URLSession(configuration: configuration)
    }
    let stream = AsyncStream<HermesChatClientEvent>.makeStream(
      bufferingPolicy: .bufferingNewest(1_000)
    )
    events = stream.stream
    continuation = stream.continuation
  }

  deinit {
    handshakeTask?.cancel()
    continuation.finish()
    socket?.cancel(with: .goingAway, reason: nil)
  }

  public func connect(
    ticket: HermesChatMultiplexTicket,
    profileConfigurations: [HermesChatClientProfile]
  ) {
    closeCurrentSocket(reason: "macOS chat reconnecting")
    let configured = Dictionary(
      uniqueKeysWithValues: profileConfigurations.map {
        (
          $0.spaceID,
          ProfileState(
            crypto: $0.crypto,
            lastSequence: $0.lastSequence,
            resumeSinceMilliseconds: $0.resumeSinceMilliseconds
          )
        )
      }
    )
    let ticketSpaces = Set(ticket.spaceIds)
    guard !configured.isEmpty, Set(configured.keys) == ticketSpaces else {
      continuation.yield(.disconnected(reason: "Hermes profile ticket 与本机密钥不匹配。"))
      return
    }

    profiles = configured
    closing = false
    ready = false
    generation += 1
    let currentGeneration = generation
    var request = URLRequest(url: relayURL)
    request.timeoutInterval = 30
    request.setValue("Bearer \(ticket.ticket)", forHTTPHeaderField: "Authorization")
    request.setValue("client", forHTTPHeaderField: "X-Hermes-Role")
    request.setValue("multiplex", forHTTPHeaderField: "X-Hermes-Mode")
    request.setValue("My-Notes-macOS/1.10", forHTTPHeaderField: "User-Agent")
    let task = session.webSocketTask(with: request)
    socket = task
    task.resume()
    receiveTask = Task { [weak self] in
      await self?.receiveLoop(socket: task, generation: currentGeneration)
    }
    pingTask = Task { [weak self] in
      await self?.pingLoop(socket: task, generation: currentGeneration)
    }
    handshakeTask = Self.makeHandshakeWatchdog(timeout: handshakeTimeout) { [weak self] in
      await self?.failHandshakeIfPending(socket: task, generation: currentGeneration)
    }
  }

  public func disconnect() {
    closing = true
    generation += 1
    closeCurrentSocket(reason: "macOS chat closed")
    profiles = [:]
  }

  public func isConnected() -> Bool { ready && socket != nil && !closing }

  public func sendMessage(
    spaceID: String,
    text: String,
    attachments: [HermesChatAttachmentDescriptor]
  ) async throws -> HermesChatMessage {
    let profile = try requireReadyProfile(spaceID)
    let id = UUID().uuidString.lowercased()
    let sentAt = Int64(Date().timeIntervalSince1970 * 1_000)
    let envelope = try profile.crypto.encryptMessageEnvelope(
      sender: "client",
      text: text,
      attachments: attachments,
      messageID: id,
      sentAt: sentAt
    )
    try await sendRouted(
      spaceID: spaceID,
      frame: try Self.outgoingMessageFrame(envelope)
    )
    return HermesChatMessage(
      id: id,
      sender: "client",
      sentAt: sentAt,
      sequence: 0,
      text: text,
      attachments: attachments
    )
  }

  public func sendInteractionAction(
    spaceID: String,
    promptID: String,
    optionID: String
  ) async throws -> String {
    let profile = try requireReadyProfile(spaceID)
    let id = UUID().uuidString.lowercased()
    let envelope = try profile.crypto.encryptInteractionEnvelope(
      promptID: promptID,
      optionID: optionID,
      messageID: id,
      sentAt: Int64(Date().timeIntervalSince1970 * 1_000)
    )
    try await sendRouted(spaceID: spaceID, frame: [
      "v": 1,
      "type": "action",
      "message": try Self.jsonObject(envelope),
    ])
    return id
  }

  public func sendTyping(spaceID: String, active: Bool) async {
    guard ready, var profile = profiles[spaceID], profile.lastTypingActive != active else { return }
    do {
      try await sendRouted(spaceID: spaceID, frame: [
        "v": 1,
        "type": "typing",
        "active": active,
      ])
      profile.lastTypingActive = active
      profiles[spaceID] = profile
    } catch {
      // Typing state is best-effort and must not interrupt message entry.
    }
  }

  private func receiveLoop(socket: URLSessionWebSocketTask, generation: Int) async {
    do {
      while !Task.isCancelled, generation == self.generation {
        let value = try await socket.receive()
        let data: Data
        switch value {
        case .data(let received): data = received
        case .string(let received): data = Data(received.utf8)
        @unknown default: continue
        }
        try await handleIncoming(data, generation: generation)
      }
    } catch {
      guard generation == self.generation, !closing else { return }
      handshakeTask?.cancel()
      handshakeTask = nil
      pingTask?.cancel()
      pingTask = nil
      ready = false
      self.socket = nil
      continuation.yield(.disconnected(reason: Self.errorMessage(error)))
    }
  }

  private func pingLoop(socket: URLSessionWebSocketTask, generation: Int) async {
    while !Task.isCancelled, generation == self.generation {
      do {
        try await Task.sleep(for: .seconds(25))
        guard generation == self.generation, !closing else { return }
        try await Self.sendPing(socket)
      } catch {
        if generation == self.generation, !closing {
          socket.cancel(with: .goingAway, reason: nil)
        }
        return
      }
    }
  }

  private func handleIncoming(_ data: Data, generation: Int) async throws {
    guard generation == self.generation,
      let frame = try JSONSerialization.jsonObject(with: data) as? [String: Any],
      let type = frame["type"] as? String
    else { throw HermesChatCryptoError.invalidMessage }

    if type == "ready" {
      guard frame["multiplex"] as? Bool == true else {
        throw HermesChatCryptoError.invalidMessage
      }
      for spaceID in profiles.keys.sorted() { try await sendResume(spaceID: spaceID) }
      handshakeTask?.cancel()
      handshakeTask = nil
      ready = true
      continuation.yield(.ready)
      return
    }
    if type == "pong" { return }
    guard let spaceID = frame["spaceId"] as? String, profiles[spaceID] != nil else {
      throw HermesChatCryptoError.invalidMessage
    }

    switch type {
    case "message", "edit":
      guard let rawEnvelope = frame["message"] else {
        throw HermesChatCryptoError.invalidMessage
      }
      let envelopeData = try JSONSerialization.data(withJSONObject: rawEnvelope)
      let envelope = try JSONDecoder().decode(HermesChatMessageEnvelope.self, from: envelopeData)
      let sequence = Self.int64(frame["seq"]) ?? 0
      var profile = try requireProfile(spaceID)
      let message = try profile.crypto.decryptMessageEnvelope(envelope, sequence: sequence)
      if type == "edit" {
        guard let target = Self.int64(frame["targetSeq"]),
          let final = frame["final"] as? Bool,
          message.replaceSequence == target,
          message.finalUpdate == final
        else { throw HermesChatCryptoError.invalidMessage }
      }
      if sequence > 0 {
        profile.lastSequence = max(profile.lastSequence, sequence)
        profiles[spaceID] = profile
      }
      continuation.yield(.message(spaceID: spaceID, message: message))
    case "resume_complete":
      if frame["hasMore"] as? Bool == true {
        try await sendResume(spaceID: spaceID)
      } else {
        let latestSequence = max(0, Self.int64(frame["latestSeq"]) ?? 0)
        if var profile = profiles[spaceID] {
          profile.lastSequence = max(profile.lastSequence, latestSequence)
          profile.resumeSinceMilliseconds = nil
          profiles[spaceID] = profile
        }
        continuation.yield(.resumed(spaceID: spaceID, latestSequence: latestSequence))
      }
    case "ack":
      let durable = frame["durable"] as? Bool ?? true
      let id = frame["id"] as? String ?? ""
      if durable {
        let sequence = Self.int64(frame["seq"]) ?? 0
        if sequence > 0, var profile = profiles[spaceID] {
          profile.lastSequence = max(profile.lastSequence, sequence)
          profiles[spaceID] = profile
        }
        continuation.yield(
          .acknowledged(spaceID: spaceID, messageID: id, sequence: sequence)
        )
      } else {
        continuation.yield(
          .actionAcknowledged(
            spaceID: spaceID,
            actionID: id,
            delivered: frame["delivered"] as? Bool ?? false
          )
        )
      }
    case "typing":
      if frame["sender"] as? String == "agent" {
        continuation.yield(
          .typing(spaceID: spaceID, active: frame["active"] as? Bool ?? false)
        )
      }
    case "error":
      continuation.yield(
        .profileError(
          spaceID: spaceID,
          reason: frame["message"] as? String ?? "Cloudflare 拒绝了聊天消息。"
        )
      )
    default:
      throw HermesChatCryptoError.invalidMessage
    }
  }

  private func sendResume(spaceID: String) async throws {
    let profile = try requireProfile(spaceID)
    try await sendRouted(
      spaceID: spaceID,
      frame: Self.resumeFrame(
        afterSequence: profile.lastSequence,
        sinceMilliseconds: profile.resumeSinceMilliseconds
      )
    )
  }

  static func resumeFrame(
    afterSequence: Int64,
    sinceMilliseconds: Int64? = nil
  ) -> [String: Any] {
    var frame: [String: Any] = [
      "v": 1,
      "type": "resume",
      "afterSeq": max(0, afterSequence),
    ]
    if let sinceMilliseconds, sinceMilliseconds > 0 {
      frame["since"] = sinceMilliseconds
    }
    return frame
  }

  private func sendRouted(spaceID: String, frame: [String: Any]) async throws {
    guard let socket, !closing else { throw URLError(.notConnectedToInternet) }
    var routed = frame
    routed["spaceId"] = spaceID
    let data = try JSONSerialization.data(withJSONObject: routed)
    guard let text = String(data: data, encoding: .utf8) else {
      throw HermesChatCryptoError.invalidMessage
    }
    try await socket.send(.string(text))
  }

  private func requireReadyProfile(_ spaceID: String) throws -> ProfileState {
    guard ready, socket != nil, !closing else { throw URLError(.notConnectedToInternet) }
    return try requireProfile(spaceID)
  }

  private func requireProfile(_ spaceID: String) throws -> ProfileState {
    guard let profile = profiles[spaceID] else { throw HermesChatCryptoError.invalidKey }
    return profile
  }

  private func closeCurrentSocket(reason: String) {
    receiveTask?.cancel()
    pingTask?.cancel()
    handshakeTask?.cancel()
    receiveTask = nil
    pingTask = nil
    handshakeTask = nil
    ready = false
    let old = socket
    socket = nil
    old?.cancel(with: .normalClosure, reason: Data(reason.utf8))
  }

  private func failHandshakeIfPending(
    socket candidate: URLSessionWebSocketTask,
    generation attemptGeneration: Int
  ) {
    guard attemptGeneration == generation, !closing, !ready,
      let currentSocket = socket, currentSocket === candidate
    else { return }
    generation += 1
    receiveTask?.cancel()
    pingTask?.cancel()
    receiveTask = nil
    pingTask = nil
    handshakeTask = nil
    ready = false
    socket = nil
    currentSocket.cancel(
      with: .goingAway,
      reason: Data("macOS chat handshake timed out".utf8)
    )
    continuation.yield(.disconnected(reason: Self.handshakeTimeoutMessage))
  }

  private static func jsonObject<T: Encodable>(_ value: T) throws -> Any {
    try JSONSerialization.jsonObject(with: JSONEncoder().encode(value))
  }

  static func outgoingMessageFrame(
    _ envelope: HermesChatMessageEnvelope
  ) throws -> [String: Any] {
    guard let frame = try jsonObject(envelope) as? [String: Any] else {
      throw HermesChatCryptoError.invalidMessage
    }
    return frame
  }

  private static func int64(_ value: Any?) -> Int64? {
    if let value = value as? Int64 { return value }
    if let value = value as? Int { return Int64(value) }
    if let value = value as? NSNumber { return value.int64Value }
    return nil
  }

  private static func sendPing(_ socket: URLSessionWebSocketTask) async throws {
    try await awaitSinglePingResult { completion in
      socket.sendPing(pongReceiveHandler: completion)
    }
  }

  /// URLSession may deliver a late ping callback while a WebSocket is being
  /// cancelled or replaced. Treat the callback as one-shot so that a duplicate
  /// delegate delivery cannot resume a checked continuation twice and crash the
  /// process.
  static func awaitSinglePingResult(
    _ start: (@escaping @Sendable (Error?) -> Void) -> Void
  ) async throws {
    try await withCheckedThrowingContinuation {
      (continuation: CheckedContinuation<Void, Error>) in
      let completion = PingCompletionGate(continuation: continuation)
      start { error in
        completion.resume(error: error)
      }
    }
  }

  static func makeHandshakeWatchdog(
    timeout: Duration,
    onTimeout: @escaping @Sendable () async -> Void
  ) -> Task<Void, Never> {
    Task {
      do {
        try await Task.sleep(for: timeout)
      } catch {
        return
      }
      guard !Task.isCancelled else { return }
      await onTimeout()
    }
  }

  private static func errorMessage(_ error: Error) -> String {
    if let localized = error as? LocalizedError, let message = localized.errorDescription {
      return message
    }
    let value = error.localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines)
    return value.isEmpty ? "Hermes WebSocket 已断开。" : value
  }

  private struct ProfileState: Sendable {
    let crypto: HermesChatCrypto
    var lastSequence: Int64
    var lastTypingActive: Bool?
    var resumeSinceMilliseconds: Int64?

    init(
      crypto: HermesChatCrypto,
      lastSequence: Int64,
      resumeSinceMilliseconds: Int64?
    ) {
      self.crypto = crypto
      self.lastSequence = max(0, lastSequence)
      self.resumeSinceMilliseconds = resumeSinceMilliseconds
    }
  }

  private final class PingCompletionGate: @unchecked Sendable {
    private let lock = NSLock()
    private var continuation: CheckedContinuation<Void, Error>?

    init(continuation: CheckedContinuation<Void, Error>) {
      self.continuation = continuation
    }

    func resume(error: Error?) {
      lock.lock()
      guard let continuation else {
        lock.unlock()
        return
      }
      self.continuation = nil
      lock.unlock()

      if let error {
        continuation.resume(throwing: error)
      } else {
        continuation.resume()
      }
    }
  }
}
